apps/web/vibes/soul/primitives/dynamic-form/index.tsx

```diff
@@ -0,0 +1,351 @@
+import { Input } from '@/vibes/soul/form/input';
+import { NumberInput } from '@/vibes/soul/form/number-input';
+import { RadioGroup } from '@/vibes/soul/form/radio-group';
+import { Select } from '@/vibes/soul/form/select';
+import { SwatchRadioGroup } from '@/vibes/soul/form/swatch-radio-group';
+import { Button } from '@/vibes/soul/primitives/button';
+
+import { Field, FieldGroup, schema } from './schema';
+
+type Action<S, P> = (state: Awaited<S>, payload: P) => S | Promise<S>;
+
+interface State<F extends Field> {
+  fields: (F | FieldGroup<F>)[];
+  lastResult: SubmissionResult | null;
+}
+
+export type DynamicFormAction<F extends Field> = Action<State<F>, FormData>;
+
+interface Props<F extends Field> {
+  fields: (F | FieldGroup<F>)[];
+  action: DynamicFormAction<F>;
+  submitLabel?: string;
+  submitName?: string;
+  submitValue?: string;
+}
+
+export function DynamicForm<F extends Field>({
+  action,
+  fields: defaultFields,
+  submitLabel = 'Submit',
+  submitName,
+  submitValue,
+}: Props<F>) {
+  const [{ lastResult, fields }, formAction] = useActionState(action, {
+    fields: defaultFields,
+    lastResult: null,
+  });
+  const dynamicSchema = schema(fields);
+  const defaultValue = fields
+    .flatMap((f) => (Array.isArray(f) ? f : [f]))
+    .reduce<z.infer<typeof dynamicSchema>>(
+      (acc, field) => ({
+        ...acc,
+        [field.name]: 'defaultValue' in field ? field.defaultValue : '',
+      }),
+      { quantity: 1 },
+    );
+  const [form, formFields] = useForm({
+    lastResult,
+    constraint: getZodConstraint(dynamicSchema),
+    onValidate({ formData }) {
+      return parseWithZod(formData, { schema: dynamicSchema });
+    },
+    defaultValue,
+    shouldValidate: 'onSubmit',
+    shouldRevalidate: 'onInput',
+    onSubmit(event, { formData }) {
+      event.preventDefault();
+
+      startTransition(() => {
+        formAction(formData);
+      });
+    },
+  });
+
+  return (
+    <FormProvider context={form.context}>
+      <form {...getFormProps(form)} action={formAction}>
+        <div className="space-y-6">
+          {fields.map((field, index) => {
+            if (Array.isArray(field)) {
+              return (
+                <div className="flex gap-4" key={index}>
+                  {field.map((f) => {
+                    const groupFormField = formFields[f.name];
+
+                    if (!groupFormField) return null;
+
+                    return (
+                      <DynamicFormField
+                        field={f}
+                        formField={groupFormField}
+                        key={groupFormField.id}
+                      />
+                    );
+                  })}
+                </div>
+              );
+            }
+
+            const formField = formFields[field.name];
+
+            if (formField == null) return null;
+
+            return <DynamicFormField field={field} formField={formField} key={field.name} />;
```

Comment:
Fo we not have a `formField.id` for these like we have above?

Reply:
Not sure I understand this question

Reply:
Meant to say "Do we not". As in, we're using `field.name` as the React key, but above we have an ID that we can use. I was wondering if conform provided us with IDs for fields, if we could use that here as well